wire-gson-support/src/main/java/com/squareup/wire/MessageTypeAdapter.java

```diff
@@ -13,161 +13,149 @@
 
-    in.beginObject();
-    while (in.peek() != JsonToken.END_OBJECT) {
-      String name = in.nextName();
+    `in`.beginObject()
+    while (`in`.peek() != JsonToken.END_OBJECT) {
+      val name = `in`.nextName()
 
-      FieldBinding<M, B> fieldBinding = fieldBindings.get(name);
+      val fieldBinding = fieldBindings[name]
       if (fieldBinding == null) {
-        in.skipValue();
+        `in`.skipValue()
       } else {
-        JsonElement element = elementAdapter.read(in);
-        Object value = parseValue(fieldBinding, element);
-        fieldBinding.set(builder, value);
+        val element = elementAdapter.read(`in`)
+        val value = parseValue(fieldBinding, element)
+        fieldBinding.set(builder, value)
       }
     }
 
-    in.endObject();
-    return builder.build();
+    `in`.endObject()
+    return builder.build()
   }
 
-  private Object parseValue(FieldBinding<?, ?> fieldBinding, JsonElement element) {
-    if (fieldBinding.label.isRepeated()) {
-      if (element.isJsonNull()) {
-        return Collections.emptyList();
+  private fun parseValue(fieldBinding: FieldBinding<*, *>, element: JsonElement): Any {
+    if (fieldBinding.label.isRepeated) {
+      if (element.isJsonNull) {
+        return emptyList<Any>()
       }
 
-      Class<?> itemType = fieldBinding.singleAdapter().javaType;
+      val itemType = fieldBinding.singleAdapter().javaType
 
-      JsonArray array = element.getAsJsonArray();
-      List<Object> result = new ArrayList<>(array.size());
-      for (JsonElement item : array) {
-        result.add(gson.fromJson(item, itemType));
+      val array = element.asJsonArray
+      val result = ArrayList<Any>(array.size())
+      for (item in array) {
+        result.add(gson.fromJson(item, itemType))
       }
-      return result;
+      return result
     }
 
     if (fieldBinding.isMap()) {
-      if (element.isJsonNull()) {
-        return Collections.emptyMap();
+      if (element.isJsonNull) {
+        return emptyMap<Any, Any>()
       }
 
-      Class<?> keyType = fieldBinding.keyAdapter().javaType;
-      Class<?> valueType = fieldBinding.singleAdapter().javaType;
+      val keyType = fieldBinding.keyAdapter().javaType
+      val valueType = fieldBinding.singleAdapter().javaType
 
-      JsonObject object = element.getAsJsonObject();
-      Map<Object, Object> result = new LinkedHashMap<>(object.size());
-      for (Map.Entry<String, JsonElement> entry : object.entrySet()) {
-        Object key = gson.fromJson(entry.getKey(), keyType);
-        Object value = gson.fromJson(entry.getValue(), valueType);
-        result.put(key, value);
+      val `object` = element.asJsonObject
+      val result = LinkedHashMap<Any, Any>(`object`.size())
+      for (entry in `object`.entrySet()) {
+        val key = gson.fromJson(entry.key, keyType)
+        val value = gson.fromJson(entry.value, valueType)
+        result[key] = value
       }
-      return result;
+      return result
```

Comment:
similar to my comment above, there's probably a way to "map" this into a "Map" (see what I did there?)

Reply:
Fixed